src/global/project/manifest.rs

```diff
@@ -49,40 +42,244 @@ impl Manifest {
             contents
                 .parse::<DocumentMut>()
                 .map(|doc| (manifest, doc))
-                .map_err(ManifestError::from)
+                .map_err(TomlError::from)
         }) {
             Ok(result) => result,
             Err(e) => e.to_fancy(MANIFEST_DEFAULT_NAME, &contents)?,
         };
 
-        let source = ManifestSource(document);
         let manifest = Self {
             path: manifest_path.to_path_buf(),
-            contents,
-            document: source,
+            document: TomlManifest::new(document),
             parsed: manifest,
         };
 
         Ok(manifest)
     }
 
-    /// Adds an environment to the project.
-    pub fn add_environment(&mut self, _name: String) -> miette::Result<()> {
-        todo!()
+    pub fn add_exposed_mapping(
+        &mut self,
+        env_name: &EnvironmentName,
+        mapping: &Mapping,
+    ) -> miette::Result<()> {
+        self.parsed
+            .envs
+            .entry(env_name.clone())
+            .or_default()
+            .exposed
+            .insert(
+                mapping.exposed_name.clone(),
+                mapping.executable_name.clone(),
+            );
+
+        self.document
+            .get_or_insert_nested_table(&format!("envs.{env_name}.exposed"))?
+            .insert(
+                &mapping.exposed_name.to_string(),
+                Item::Value(toml_edit::Value::from(mapping.executable_name.clone())),
+            );
+
+        tracing::debug!("Added exposed mapping {mapping} to toml document");
+        Ok(())
     }
 
-    /// Removes an environment from the project.
-    pub fn remove_environment(&mut self, _name: &str) -> miette::Result<bool> {
-        todo!()
+    pub fn remove_exposed_name(
+        &mut self,
+        env_name: &EnvironmentName,
+        exposed_name: &ExposedName,
+    ) -> miette::Result<()> {
+        self.parsed
+            .envs
+            .get_mut(env_name)
+            .ok_or_else(|| miette::miette!("[envs.{env_name}] needs to exist"))?
+            .exposed
+            .shift_remove(exposed_name);
+
+        self.document
+            .get_or_insert_nested_table(&format!("envs.{env_name}.exposed"))?
+            .remove(&exposed_name.to_string())
+            .ok_or_else(|| miette::miette!("The exposed name {exposed_name} doesn't exist"))?;
+
+        tracing::debug!("Removed exposed mapping {exposed_name} from toml document");
+        Ok(())
     }
 
-    /// Add a matchspec to the manifest
-    pub fn add_dependency(&mut self, _spec: &MatchSpec) -> miette::Result<bool> {
-        todo!()
+    /// Save the manifest to the file and update the parsed_manifest
+    pub async fn save(&mut self) -> miette::Result<()> {
+        let contents = self.document.to_string();
+        self.parsed = ParsedManifest::from_toml_str(&contents)?;
```

Comment:
Is this still needed now that you modify it directly?

Reply:
No, it isn't. Nice find!